Improve SUBSCRIBE documentation for AS OF and PROGRESS (#36904)
MaterializeInc/materialize: doc/user/content/sql/subscribe.md

The value in the `AS OF` clause is automatically [cast to `mz_timestamp`](../../sql/types/mz_timestamp/#valid-casts) with an assignment or implicit cast.

`AS OF AT LEAST <timestamp>` is a _lower bound_ rather than an exact start:
Materialize begins at the timestamp you specify, or later if that timestamp is no

> longer available. Plain `AS OF` instead requires the exact timestamp and returns an

Reviewer: Plain \`AS OF\` -> \`AS OF \<timestamp\>\`

### `PROGRESS`

> The `PROGRESS` option (`WITH (PROGRESS)`) adds an `mz_progressed` column that lets
> you tell the difference between "nothing has changed yet" and "Materialize is
> stalled." It changes the output in two ways:
> 
>   * An additional `mz_progressed` column appears. When it is `false`, the row is a
>     normal update. When it is `true`, the row is a **progress message**: every
>     column except `mz_timestamp` is `NULL` and should be ignored. The row exists
>     only to report that time has advanced.
> 
>   * The first row emitted is always a progress message at the subscribe's
>     [`AS OF`](#as-of) timestamp.

Reviewer: maybe ?
```suggestion
You can use the `WITH (PROGRESS)` option to add progress messages to the output.
Including progress messages allows you to distinguish between a legitimate
period of no updates and a stall in Materialize.

When specified, the `PROGRESS` option adds an `mz_progressed` column to each
row.

- When `mz_progressed` is `true`, the row is a **progress message**; every
  column except `mz_timestamp` and `mz_progressed` is `NULL` and should be
  ignored. The row exists only to report that time has advanced.

- When `mz_progressed` is `false`, the row is a normal update.

The first row emitted is always a progress message at the `SUBSCRIBE`'s [`AS
OF`](#as-of) timestamp.
```

#### What a progress message guarantees

> {{< note >}}
> A progress message at `mz_timestamp = X` means: the `SUBSCRIBE` will never again
> emit an update at a time **strictly less than** X. It does **not** mean time X is
> finished—more updates can still arrive at exactly X.
> {{< /note >}}
> 
> In other words, when you see a progress message at X, every timestamp _before_ X
> is now complete and safe to process. Timestamp X itself is only complete once you
> see a _later_ progress message (or any update at a time greater than X), since
> [`mz_timestamp` is always non-decreasing](#output).
> 
> The same reasoning applies to ordinary updates: because timestamps never decrease,
> the moment you see _any_ row (a progress message or a data update) at timestamp X,
> all updates at times strictly less than X have already been delivered.
> 
> A common misconception is that a progress message at X seals timestamp X. It does
> not—it seals everything _before_ X. This is why, when resuming a subscription, you
> restart `AS OF <last_progress_timestamp> - 1`: timestamp `last_progress_timestamp`
> was never sealed, so you must re-request it. See [Durable
> subscriptions](/transform-data/patterns/durable-subscriptions/).

Reviewer: to better flow with the heading and cut down a little bit of the repetition:

```suggestion
A progress message (`mz_progressed = true`) at `mz_timestamp = X` guarantees
that the `SUBSCRIBE` will never again emit an update with a timestamp **strictly
less than** X. That is, when you see a progress message with timestamp X, every
timestamp _before_ X is now complete or *sealed*.

{{< note >}}

A progress message at X does **not** mean timestamp X is sealed; more updates
can still arrive with `mz_timestamp = X`.

{{< /note >}}

Timestamp X itself is only sealed once you see a _later_ progress message (or
any update at a timestamp greater than X), since [`mz_timestamp` is always
non-decreasing](#output). The same applies to ordinary updates. Because
timestamps never decrease, once you see _any_ row (a progress message or a data
update) with timestamp X, all updates at timestamps strictly less than X have
already been delivered.

As such, when resuming a subscription, you restart with `AS OF
<last_progress_timestamp> - 1`: timestamp `last_progress_timestamp` was never
sealed, so you must re-request it. See [Durable
subscriptions](/transform-data/patterns/durable-subscriptions/).
```

#### Progress messages are not emitted for every timestamp

> Materialize only emits an explicit progress message when it would otherwise have

Reviewer: ? except for the first row?

nothing to say for a while; ordinary updates already imply progress. The following
is a valid sequence of updates:

```nofmt
mz_timestamp | mz_progressed | mz_diff | column1
-------------|---------------|---------|---------------

> 1            | false         | 1       | data            <- timestamp 1 update

Reviewer: Just wondering .. above, we mention that the first message is always a progress message ... whereas .. here it's an update message.

2            | false         | 1       | more data       <- proves timestamp 1 is complete
3            | false         | 1       | even more data  <- proves timestamp 2 is complete
4            | true          | NULL    | NULL            <- proves timestamp 3 is complete; 4 may still get data
```

Notice that there are no explicit progress messages for timestamps `1`, `2`, or
`3`. The update at timestamp `2` already implies that there are no more updates for
timestamp `1` (timestamps are always presented in non-decreasing order), and so
on. The explicit progress message at timestamp `4` implies that there are no more
updates for either timestamp `2` or `3`—but that there may be more data arriving at
timestamp `4`.

> #### Using progress messages

Reviewer: ```suggestion
#### Processing data
```

* If [`PROGRESS`](#progress) is set, Materialize also returns the `mz_progressed`
column. Each progress row will have a `NULL` key and a `NULL` value.

> * Within each timestamp, the rows are the _net_ result of all changes to each key

Reviewer: to each key -> for each key ?
